### ISSUE TITLE
FlatList onEndReached , onViewableItemsChanged ...etc wont trigger occasionally & workaround

### ISSUE THREAD

**Issue by yingxuepeng** [facebook/react-native#30456]
**TL;DR**
viewabilityConfig={{ itemVisiblePercentThreshold: 100 }} &&
a float width both FlatList  and PageSize( or height) &&
PagingEnabled 
**may causing a loss of precision problem**

## Description
My project use FlatList as a HorizontalPageView ( each child have the same size as FlatList ), then onViewableItemsChanged triggered correctly on PORTRAIT mode( both init and after scrollToIndex will be called).
But when the test team use some device (such as huawei mateXX) in LANDSCAPE mode, it wont trigger anymore.

I found that in LANDSCAPE mode, the width of <FlatList> and getItemLayout returned width is "719.66666870117188". 
When I set these width to a Integer value, it works 100% correct. 
So it is highly because the "if(...)" statement  to trigger "onViewableItemsChanged" failed with a float precision problem.

## Workaround
1.viewabilityConfig={{ itemVisiblePercentThreshold: _99_}} : but may still have problem when FlatList is extremely long, i think.
2.fixed Math.floor(width): add a new state.flatListSize, Move FlatList into  View, set onLayout callback within this view, set the flatListSize when View.onLayout is called. only render FlatList when flatListSize  is not undefined, and with the Integer width & height style override. (<---My Code now)

## Suggestion
FlatList may add a new parameter to change self's width & height to Integer(Math.floor may be better), then getItemLayout return the flatList's width & height,  we can return the page layout equal the FlatList size **exactly**
All View Component with threshold should consider precision problem.

## Reproduce code example
```TSX
export class TestScreen extends React.Component<TestScreenProps, TestScreenState> {
  static readonly Name = 'TestScreen';
  private curIndex = 0;
  private flatListRef = React.createRef<FlatList>();
  public constructor(props) {
    super(props);
    // Orientation.lockToLandscape();
  }
  componentDidMount() {}

  render() {
    // let width = 700 / 4;
    let width = 700 / 3;
    let height = 500;
    return (
      <View>
        <FlatList
          style={{ width, height }}
          ref={this.flatListRef}
          // data
          data={[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
          // paging
          initialScrollIndex={0}
          initialNumToRender={0}
          windowSize={2}
          horizontal={true}
          pagingEnabled={true}
          scrollEnabled={false}
          showsHorizontalScrollIndicator={false}
          renderItem={info => {
            return (
              <View style={{ width, height, backgroundColor: '#ff00ff' }}>
                <Text style={{ fontSize: 40 }}>{info.item}</Text>
              </View>
            );
          }}
          keyExtractor={(item, index) => index + ''}
          onViewableItemsChanged={this.onViewableItemsChanged}
          viewabilityConfig={{
            itemVisiblePercentThreshold: 100
          }}
          getItemLayout={(data, index) => {
            return {
              length: width,
              offset: index * width,
              index
            };
          }}
        />
        <TouchableHighlight
          style={{ padding: 20 }}
          onPress={() => {
            this.flatListRef.current?.scrollToIndex({ index: this.curIndex++, animated: true });
          }}
        >
          <Text>Next</Text>
        </TouchableHighlight>
      </View>
    );
  }
  public onViewableItemsChanged = ({ viewableItems, changed }) => {
    console.log('TestScreen onViewableItemsChanged', viewableItems.length > 0 ? viewableItems[0].index : 'Empty');
  };
}
```

## React Native version:
System:
    OS: Windows 10 10.0.18363
    CPU: (12) x64 AMD Ryzen 5 3600X 6-Core Processor
    Memory: 4.81 GB / 15.95 GB
  Binaries:
    Node: 12.18.3 - F:\software\nodejs\node.EXE
    Yarn: Not Found
    npm: 6.14.7 - F:\software\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.17763.0, 10.0.18362.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.6.30225.117 (Visual Studio Community 2019), 15.9.28307.1064 (Visual Studio Community 2017)
  Languages:
    Java: 1.8.0_251
    Python: Not Found
  npmPackages:
    @react-native-community/cli: Not Found
    react: ^16.11.0 => 16.11.0
    react-native: ^0.62.2 => 0.62.2
    react-native-windows: Not Found
  npmGlobalPackages:
    *react-native*: Not Found

#25656  & some onViewableItemChanged problem
